### PR TITLE
Added Cassandra Support

### PR DESCRIPTION
**Pull request by eschizoid**

### CODE REVIEW

File: optimus/io/jdbc.py [hunk 1]
```diff
@@ -278,20 +299,20 @@ def df_to_table(self, df, table, mode="overwrite"):
         conf.save()
 
     @staticmethod
-    def _limit(limit=None):
+    def _limit(df: DataFrame, limit=None) -> DataFrame:
```
Comment (arpit1997) [inline optimus/io/jdbc.py]: Type annotations are a really good idea, Can you make a separate PR for the complete codebase? I am assuming the IDEs/other tools can do this automatically for about 75% of the codebase accurately. Let me know your thoughts

File: optimus/io/jdbc.py [hunk 2]
```diff
@@ -227,31 +244,35 @@ def execute(self, query, limit=None):
...
+            query = "(" + query + ") AS t"
 
         logger.print(query)
         logger.print(self.url)
 
         conf = Spark.instance.spark.read \
-            .format("jdbc") \
+            .format("jdbc" if not DriverResolver.CASSANDRA.__str__() else "org.apache.spark.sql.cassandra") \
             .option("url", self.url) \
-            .option("dbtable", query) \
```
Comment (arpit1997) [inline optimus/io/jdbc.py]: @eschizoid Any reason for not supplying the query here? I can't see the query specified in your code. `execute` method receives arbitrary queries so this might work for loading the complete data (or with some limit) it won't work with  arbitrary queries.

File: optimus/io/jdbc.py [hunk 3]
```diff
@@ -227,31 +244,35 @@ def execute(self, query, limit=None):
...
+        if self.db_driver == DriverResolver.ORACLE.__str__() \
+                or self.db_driver == DriverResolver.POSTGRES_SQL.__str__() \
+                or self.db_driver == DriverResolver.PRESTO.__str__():
             conf.option("driver", self.driver_option)
 
-        return conf.load()
+        if self.db_driver == DriverResolver.CASSANDRA.__str__():
+            conf.options(table="main", keyspace=self.cassandra_keyspace)
+
+        return self._limit(conf.load(), limit)
```
Comment (arpit1997) [inline optimus/io/jdbc.py]: Any reason for adding the limit after loading the data instead of specifying in the query?

File: optimus/io/jdbc.py [hunk 4]
```diff
@@ -227,31 +244,35 @@ def execute(self, query, limit=None):
...
 
-        if self.db_driver == "oracle" or self.db_driver == 'postgresql' or self.db_driver == 'presto':
+        if self.db_driver == DriverResolver.ORACLE.__str__() \
+                or self.db_driver == DriverResolver.POSTGRES_SQL.__str__() \
+                or self.db_driver == DriverResolver.PRESTO.__str__():
             conf.option("driver", self.driver_option)
 
-        return conf.load()
+        if self.db_driver == DriverResolver.CASSANDRA.__str__():
+            conf.options(table="main", keyspace=self.cassandra_keyspace)
```
Comment (arpit1997) [inline optimus/io/jdbc.py]: Can you make the table names configurable as well like keyspace? You can use the table name here as well from JDBC class contructor

**Comment by arpit1997** [comment]
@eschizoid Thank for creating the PR. Really appreciate the changes and ideas on refactoring. I have added some comments. Please let me know if you need clarification or inputs on this.

File: optimus/io/jdbc.py [hunk 5]
```diff
@@ -15,65 +18,73 @@ class JDBC:
...
         """
 
         self.db_driver = driver
         self.oracle_sid = oracle_sid
+        self.cassandra_keyspace = cassandra_keyspace
 
         # Handle the default port
-        if self.db_driver == "redshift":
-            if port is None: self.port = 5439
+        if self.db_driver == DriverResolver.REDSHIFT.__str__():
```
[the remark this answers is not in the document]
Comment (eschizoid) [reply optimus/io/jdbc.py]: That is exactly was I was going to do and implement the strategy pattern and start adding some tests (other than running the notebook manually). However, in order to keep this PR focus in the Cassandra feature, I would like to do it on a follow-up PR. Let me know your thoughts

File: optimus/io/jdbc.py (hunk 1 above)
Comment (eschizoid) [reply optimus/io/jdbc.py]: Yup kill fix that :)

File: optimus/io/jdbc.py (hunk 2 above)
Comment (eschizoid) [reply optimus/io/jdbc.py]: @arpit1997 Good catch, this was removed by mistake will add it back

Comment (eschizoid) [reply optimus/io/jdbc.py]: Good catch, this was removed by mistake. Will add it back

File: optimus/io/jdbc.py (hunk 3 above)
Comment (eschizoid) [reply optimus/io/jdbc.py]: @arpit1997 Yeah because the `cassandra-spark-connector` doesn't support query. So I don't have another way to enforce the limit other than doing it at the `df` level.

File: optimus/io/jdbc.py (hunk 4 above)
Comment (eschizoid) [reply optimus/io/jdbc.py]: Yup will take care of it

File: optimus/io/jdbc.py (hunk 2 above)
Comment (eschizoid) [reply optimus/io/jdbc.py]: Fixed

**Comment by arpit1997** [comment]
Sorry, I had been busy. I will have to test the PR for regression once. Otherwise everything looks good.

**Comment by eschizoid** [comment]
@arpit1997 I fixed the merge conflicts, any idea when can this PR be merged down?

**Comment by arpit1997** [comment]
@eschizoid Apologies for the late reply. I tested the PR. I found some regression issues, Cassandra support was breaking other JDBC connections. I fixed the initial one I found which wasn't letting users connect to any other JDBC database (check here https://github.com/ironmussa/Optimus/pull/665/commits/8da900c3424146ecaefa88caf170e88e274211cf)

I left a couple of other comments which are causing regression. Can you test the changes you made with MySQL, Postgres, Oracle and Redshift? It would be really helpful. The docker-compose file for getting these DBs up and running are available at https://github.com/ironmussa/Optimus/blob/master/tests/sql/docker-compose.yml

Please let me know if you run into any issues

File: optimus/io/jdbc.py [hunk 6]
```diff
@@ -228,31 +246,36 @@ def execute(self, query, limit=None):
...
             conf.option("password", self.password)
 
-        if self.db_driver == "oracle" or self.db_driver == 'postgresql' or self.db_driver == 'presto':
+        if self.db_driver == DriverResolver.ORACLE.__str__() \
+                or self.db_driver == DriverResolver.POSTGRES_SQL.__str__() \
+                or self.db_driver == DriverResolver.PRESTO.__str__():
             conf.option("driver", self.driver_option)
 
-        return conf.load()
+        if self.db_driver == DriverResolver.CASSANDRA.__str__():
```
Comment (arpit1997) [inline optimus/io/jdbc.py]: @eschizoid `execute` method returning a dataframe is causing regression, its failing for `df = db_mysql.table_to_df('main', limit='all')` for a connected mysql db object `db_mysql`. See here https://github.com/ironmussa/Optimus/blob/master/tests/sql/sql.ipynb

Comment (eschizoid) [reply optimus/io/jdbc.py]: @arpit1997 Unfortunately I can't run MySQL tests since the `requirements.txt` or `requirements-test.txt` does not have the dependency. When I runt this block from the notebook:

```

host = "127.0.0.1"
port= 3306
driver = "mysql"
user = "SA"
password = "SA123"
db_name = "optimus"

db = records.Database('{DRIVER}://{USER}:{PASSWORD}@{HOST}:{PORT}/{DB}'.format(DRIVER=driver, USER=user, PASSWORD=password, HOST=host, PORT=port,DB=db_name))

db.query_file('main.sql')
db.query_file('secondary.sql')
```

I get this error:
---------------------------------------------------------------------------

ModuleNotFoundError                       Traceback (most recent call last)

<ipython-input-27-56e9fc4d31b0> in <module>
      7 db_name = "optimus"
      8 
----> 9 db = records.Database('{DRIVER}://{USER}:{PASSWORD}@{HOST}:{PORT}/{DB}'.format(DRIVER=driver, USER=user, PASSWORD=password, HOST=host, PORT=port,DB=db_name))
     10 
     11 db.query_file('main.sql')

/usr/local/lib/python3.7/site-packages/records.py in __init__(self, db_url, **kwargs)
    260 
    261         # Create an engine.
--> 262         self._engine = create_engine(self.db_url, **kwargs)
    263         self.open = True
    264 

/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/__init__.py in create_engine(*args, **kwargs)
    454     strategy = kwargs.pop("strategy", default_strategy)
    455     strategy = strategies.strategies[strategy]
--> 456     return strategy.create(*args, **kwargs)
    457 
    458 

/usr/local/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py in create(self, name_or_url, **kwargs)
     85                 if k in kwargs:
     86                     dbapi_args[k] = pop_kwarg(k)
---> 87             dbapi = dialect_cls.dbapi(**dbapi_args)
     88 
     89         dialect_args["dbapi"] = dbapi

/usr/local/lib/python3.7/site-packages/sqlalchemy/dialects/mysql/mysqldb.py in dbapi(cls)
    116     @classmethod
    117     def dbapi(cls):
--> 118         return __import__("MySQLdb")
    119 
    120     def on_connect(self):

ModuleNotFoundError: No module named 'MySQLdb'
```

I've tried to install several packages pip packages and none of them seem to work (I am using virtual env with pythbon 3.7)...

If you have any suggestions on how to fix this please legt me know

Comment (arpit1997) [reply optimus/io/jdbc.py]: MySqldb might be a lot of pain to install because of system specific dependencies. Please try running for others DBs like postgres, oracle and redshift. I will check things for MySql. 
 Let me know if you run into problems

Comment (argenisleon) [reply optimus/io/jdbc.py]: @arpit1997 @Jarrioja setup some database to test .Maybe It could help:

MySQL
Server: 165.227.196.70:3306
Username: test
Password: test
Database: optimus

PostgresSQL
Server: 165.227.196.70:5432
Username: testuser
Password: test
Database: optimus

Redshift
Server: 165.227.196.70:5439
Username: testuser
Password: test
Database: optimus

MS SQL
Server:165.227.196.70:1433
Username: test
Password: test*0261
Database: optimus

Comment (eschizoid) [reply optimus/io/jdbc.py]: @arpit1997 Sorry for my late reply works has kept me busy.

Anyways, I tested Redshift and Postgres after my last commit, and the method:

```
df = db_postgres.table_to_df('main', limit='all')
```

is no longer failing, please let me know if there is something else that needs to be done.

**Comment by eschizoid** [comment]
hey @arpit1997 wondering if you haad a chance to review this :)

**Comment by argenisleon** [comment]
@eschizoid Sorry, for the delay. I'll be reviewing this today

**Comment by argenisleon** [comment]
@eschizoid All the tests on my side are working. Merging.

**Comment by eschizoid** [comment]
@argenisleon Cool, I have a couple of follow up PR's to refactor the code and make it more testable.

**Comment by argenisleon** [comment]
@eschizoid. I squish a bug with the port param, and added support to sqlite. Also working MSSQL.
If you want to take a look here https://github.com/ironmussa/Optimus/tree/feature/sqlite_sqlserver_support

**Comment by eschizoid** [comment]
@argenisleon Awesome will take a look today :)
